packages/dx-react-scheduler/src/plugins/drag-drop-provider.jsx

```diff
@@ -0,0 +1,322 @@
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
+import moment from 'moment';
+import {
+  Plugin, Template, TemplatePlaceholder,
+  TemplateConnector, DropTarget, DragSource,
+  DragDropProvider as DragDropProviderCore,
+} from '@devexpress/dx-react-core';
+import {
+  cellIndex,
+  cellData,
+  cellType,
+  allDayRects,
+  verticalTimeTableRects,
+  horizontalTimeTableRects,
+  getAppointmentStyle,
+  intervalDuration,
+  VERTICAL_TYPE,
+  SCROLL_OFFSET,
+  SCROLL_SPEED_PX,
+  SECONDS,
+} from '@devexpress/dx-scheduler-core';
+
+const pluginDependencies = [
+  { name: 'EditingState' },
+  { name: 'Appointments' },
+  { name: 'DayView', optional: true },
+  { name: 'WeekView', optional: true },
+  { name: 'MonthView', optional: true },
+  { name: 'AllDayPanel', optional: true },
+];
+
+export class DragDropProvider extends React.PureComponent {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      startTime: null,
+      endTime: null,
+      payload: undefined,
+    };
+
+    this.resetCache = this.resetCache.bind(this);
+    this.onOver = (getters, actions) => args => this.handleOver.bind(this)(args, getters, actions);
+    this.onDrop = actions => () => this.handleDrop.bind(this)(actions);
+    this.onPayloadChange = actions => args => this.handlePayloadChange.bind(this)(args, actions);
+
+    this.timeTableRects = [];
+    this.allDayRects = [];
+    this.offsetTimeTop = null;
+    this.appointmentStartTime = null;
+    this.appointmentEndTime = null;
+  }
+
+  resetCache() {
+    this.timeTableRects = [];
+    this.allDayRects = [];
+    this.offsetTimeTop = null;
+    this.appointmentStartTime = null;
+    this.appointmentEndTime = null;
+
+    this.setState({
+      payload: null,
+      startTime: null,
+      endTime: null,
+    });
+  }
+
+  handlePayloadChange({ payload }, { commitChangedAppointment, stopEditAppointment }) {
+    if (payload) return;
+
+    const { payload: prevPayload } = this.state;
+
+    stopEditAppointment({ appointmentId: prevPayload.id });
+    commitChangedAppointment({ appointmentId: prevPayload.id });
+    this.resetCache();
+  }
+
+  handleOver(
```

Comment:
This method is too long, it needs to be refactored

Reply:
So many variables... I am afraid I can't make it small function.

Reply:
We can extract at least 4 functions - 
lines 95 - 106 
130 - 147 
149 - 175 
181 - 193